apiserver/internal/handlers/resources/download/downloader.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package download
+
+import (
+	"context"
+	"crypto/sha512"
+	"encoding/hex"
+	"io"
+	"os"
+
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/internal/errors"
+)
+
+const (
+	// ErrUnexpectedHash is returned when the hash of the downloaded resource
+	// does not match the expected hash.
+	ErrUnexpectedHash = errors.ConstError("downloaded resource has unexpected hash")
+
+	// ErrUnexpectedSize is returned when the size of the downloaded resources
+	// does not match the expected size.
+	ErrUnexpectedSize = errors.ConstError("downloaded resource has unexpected size")
+)
+
+// FileSystem defines a file system for modifying files on a users system.
+type FileSystem interface {
+	// CreateTemp creates a new temporary file in the directory dir,
+	// opens the file for reading and writing, and returns the resulting file.
+	CreateTemp(dir, pattern string) (*os.File, error)
+
+	// Open opens the named file for reading. If successful, methods on
+	// the returned file can be used for reading; the associated file
```

Comment:
todo: update comment line, partial thought?

Reply:
Partially copied from the standard lib, have I've removed the second sentance, it didn't add anything.